### ISSUE TITLE
RPI +ESP8266

### ISSUE THREAD

**Issue by Fisher92** [scottlawsonbc/audio-reactive-led-strip#25]
Buffer overflows. Poor performance/

Tried cutting the fps to 30 and freq bins to 12. Runs incredibly slow for some reason. RPI2

**Comment by scottlawsonbc** [comment]
Thanks for the report.

I think I can fix this. The audio samples are being read from the microphone in short duration chunks. The number of audio samples in each chunk depends on the FPS and the microphone sampling rate:
```
audio samples = (audio sample rate) / FPS
```
A frequency domain representation of these audio samples is obtained through a fast Fourier transform (FFT). After some testing, I discovered a surprising relationship between the number of audio samples and the FFT computation time.

The test results are shown below:

![image](https://cloud.githubusercontent.com/assets/2927085/22402326/75bec9c8-e5a7-11e6-97c6-eae83756a844.png)

This tells us that the number of audio samples significantly affects performance, and even small changes in sample size can affect performance significantly, and would explain why some people have this issue and others don't. I'll release a change tonight that forces the audio sample length to be a power of two. This should improve performance by a factor of two or three at least.

**Comment by Fisher92** [comment]
Much appreciated. Nice project btw!

**Comment by scottlawsonbc** [comment]
Replace the contents of `visualization.py` with [this version](https://github.com/scottlawsonbc/audio-reactive-led-strip/blob/master/python/visualization.py).

Does this help with the poor performance?

**Comment by Fisher92** [comment]
Seems a bit better but still crazy slow. also the fps seems to drag from 60 down to 20 over about a minute...

Maybe the bloated raspbian jesse im using is a poor choice?
![image](https://cloud.githubusercontent.com/assets/11381045/22403091/c8e21572-e658-11e6-9c5e-c45f5a24ef45.png)

**Comment by Fisher92** [comment]
more info

![image](https://cloud.githubusercontent.com/assets/11381045/22403114/95b60bee-e659-11e6-8ff3-9af670ad0db0.png)

**Comment by scottlawsonbc** [comment]
Are you using the GUI?

**Comment by Fisher92** [comment]
No - GUI and FPS turned off. tried it on rpi3 (jesse again) with the same outcome.
Config.py

"""Settings for audio reactive LED strip"""
from __future__ import print_function
from __future__ import division
import os

DEVICE = 'esp8266'
"""Device used to control LED strip. Must be 'pi',  'esp8266' or 'blinkstick'

'esp8266' means that you are using an ESP8266 module to control the LED strip
and commands will be sent to the ESP8266 over WiFi.

'pi' means that you are using a Raspberry Pi as a standalone unit to process
audio input and control the LED strip directly.

'blinkstick' means that a BlinkstickPro is connected to this PC which will be used
to control the leds connected to it.
"""

if DEVICE == 'esp8266':
    UDP_IP = '192.168.1.150'
    """IP address of the ESP8266. Must match IP in ws2812_controller.ino"""
    UDP_PORT = 7777
    """Port number used for socket communication between Python and ESP8266"""
    SOFTWARE_GAMMA_CORRECTION = False
    """Set to False because the firmware handles gamma correction + dither"""

if DEVICE == 'pi':
    LED_PIN = 18
    """GPIO pin connected to the LED strip pixels (must support PWM)"""
    LED_FREQ_HZ = 800000
    """LED signal frequency in Hz (usually 800kHz)"""
    LED_DMA = 5
    """DMA channel used for generating PWM signal (try 5)"""
    BRIGHTNESS = 255
    """Brightness of LED strip between 0 and 255"""
    LED_INVERT = True
    """Set True if using an inverting logic level converter"""
    SOFTWARE_GAMMA_CORRECTION = True
    """Set to True because Raspberry Pi doesn't use hardware dithering"""

if DEVICE == 'blinkstick':
    SOFTWARE_GAMMA_CORRECTION = True
    """Set to True because blinkstick doesn't use hardware dithering"""

USE_GUI = False
"""Whether or not to display a PyQtGraph GUI plot of visualization"""

DISPLAY_FPS = False
"""Whether to display the FPS when running (can reduce performance)"""

N_PIXELS = 174
"""Number of pixels in the LED strip (must match ESP8266 firmware)"""

GAMMA_TABLE_PATH = os.path.join(os.path.dirname(__file__), 'gamma_table.npy')
"""Location of the gamma correction table"""

MIC_RATE = 44100
"""Sampling frequency of the microphone in Hz"""

FPS = 60
"""Desired refresh rate of the visualization (frames per second)

FPS indicates the desired refresh rate, or frames-per-second, of the audio
visualization. The actual refresh rate may be lower if the computer cannot keep
up with desired FPS value.

Higher framerates improve "responsiveness" and reduce the latency of the
visualization but are more computationally expensive.

Low framerates are less computationally expensive, but the visualization may
appear "sluggish" or out of sync with the audio being played if it is too low.

The FPS should not exceed the maximum refresh rate of the LED strip, which
depends on how long the LED strip is.
"""
_max_led_FPS = int(((N_PIXELS * 30e-6) + 50e-6)**-1.0)
assert FPS <= _max_led_FPS, 'FPS must be <= {}'.format(_max_led_FPS)

MIN_FREQUENCY = 200
"""Frequencies below this value will be removed during audio processing"""

MAX_FREQUENCY = 12000
"""Frequencies above this value will be removed during audio processing"""

N_FFT_BINS = 24
"""Number of frequency bins to use when transforming audio to frequency domain

Fast Fourier transforms are used to transform time-domain audio data to the
frequency domain. The frequencies present in the audio signal are assigned
to their respective frequency bins. This value indicates the number of
frequency bins to use.

A small number of bins reduces the frequency resolution of the visualization
but improves amplitude resolution. The opposite is true when using a large
number of bins. More bins is not always better!

There is no point using more bins than there are pixels on the LED strip.
"""

N_ROLLING_HISTORY = 2
"""Number of past audio frames to include in the rolling window"""

MIN_VOLUME_THRESHOLD = 1e-7
"""No music visualization displayed if recorded audio volume below threshold"""

**Comment by scottlawsonbc** [comment]
Just to double check, is your microphone rate correct?

`MIC_RATE = 44100`

Using an incorrect microphone rate can cause performance issues. `MIC_RATE = 44100` and `MIC_RATE = 48000`  are the two most common sample rates.

**Comment by Fisher92** [comment]
I used 44100 on my pc and it worked fine- tried 48000 on the pi just to make sure and it didnt make a difference?

**Comment by scottlawsonbc** [comment]
Are you using SSH with FPS output on? Excessive printing over SSH can cause performance issues. If you are using SSH with FPS output enabled, try disabling FPS output and see if that improves performance.

**Comment by Fisher92** [comment]
Hey Scott, I had the same thought and have tried turning FPS off and also tried running locally from the PI to no avail.

-Ill try running the leds from the pis GPIO when I get home later to eliminate network issues as the cause.

**Comment by Fisher92** [comment]
PS. have you had this working using an RPI and esp8266. The documentation only has PC+esp8266 or standalone pi?

**Comment by scottlawsonbc** [comment]
> PS. have you had this working using an RPI and esp8266. The documentation only has PC+esp8266 or standalone pi?

Yep. I used my Raspberry Pi 2 to control my ESP8266. I was getting around 40 FPS stable.

I wonder if your Pi is missing an optimized BLAS package. BLAS can have an enormous impact on performance. What is the output if you command?
```
python -c 'import numpy; numpy.show_config()'
```

**Comment by Fisher92** [comment]
![image](https://cloud.githubusercontent.com/assets/11381045/22419425/26e3e152-e729-11e6-8d9a-06678a1ef343.png)

I had a go getting the pi to work standalone and it still seemed to be playing up -havent looked into into it properly yet but im wondering if it could be the sound card im using (cheapo logitech). Pi seemed to pick it up fine and it's definitely responding to the audio so im not sure...

**Comment by scottlawsonbc** [comment]
Hmm... my Pi 2 config output is the same.

I ran some tests on my Pi 2. I am now able to reproduce this issue using default settings except that I change `N_PIXELS=60` to `N_PIXELS=174`.

The Pi starts at a high FPS, but then the FPS quickly drops and buffer overflows occur multiple times per second. On my machine, it dropped from 60 FPS to 38. Over the next two minutes it continued to decrease to less than 20 FPS, throwing hundreds of buffer overflows.

Does this sound like the problem you are having?

I also tested performance after commenting out `led.update()`. Everything worked without issue when LED strip updates were disabled. This implies that the problem is related to network communication in some way.

Hmm... I wonder if this is related to the MTU size issue #20 I'll look into this further. Thank you for reporting this.
.

**Comment by Fisher92** [comment]
yes exactly
ill reduce the leds and use a stock config

**Comment by Fisher92** [comment]
definitely makes a difference. still cant handle 20leds at 30 fps though :(
What distro are you using - jesse?
![image](https://cloud.githubusercontent.com/assets/11381045/22423104/9aaebb54-e73b-11e6-8984-89e5aeafbaa0.png)

**Comment by scottlawsonbc** [comment]
With 60 LEDs and GUI disabled you should be able to get 60 FPS. At the very least you should be able to get 30 FPS.

What is your FPS when you run with 60 LEDs + GUI disabled?

**Comment by Fisher92** [comment]
The screenshot above is 30leds gui off and this one is 10leds gui off after about 10min running...
are you running jesse or one of the lighter ones?

![image](https://cloud.githubusercontent.com/assets/11381045/22424146/df1ac990-e740-11e6-963d-dc00436570e5.png)

**Comment by scottlawsonbc** [comment]
I'm running Jessie lite.

I've optimized the visualization as much as I can. I've just added around a dozen performance optimizations. I've also reduced the frequency that information is printed to screen, which will help SSH users.

I can achieve a solid 60 FPS with 60 pixels, but I barely manage to get 24 FPS when using 174 pixels. When I comment out `led.update()` everything runs very fast so I suspect that the bottleneck is the UDP transfer rate.

Performance wise, this is the best I can do until I release the next version. I was hoping to release by the end of January but it might be another week or so.
